### PR TITLE
Refactor Integer.Format.Decimal to avoid internal array allocation.

### PR DESCRIPTION
**Pull request by jemc**
This new approach uses some bit magic to closely estimate digit count, and a lookup table to quickly get the divisor for a given digit count.

### CODE REVIEW

File: core/Integer.Format.savi [hunk 1]
```diff
@@ -42,17 +42,28 @@
...
 
-    // Collect the digits we need to print, from least to most significant.
-    // TODO: Avoid this temporary array allocation here.
-    digits Array(U8) = []
-    while (value > zero) (
-      digits << (value % ten).u8 + '0'
-      value = value / ten
+    // Convert the value to a U64 so we don't have to specialize code by size.
+    // TODO: It would likely be slightly faster to specialize by size here.
+    u64 = value.u64
```
Comment (mneumann) [inline core/Integer.Format.savi:47]: this would break for `U128`

Comment (jemc) [reply core/Integer.Format.savi:47]: True, although we don't have `U128` at this time, and there are lots of other places that will also break when/if we add `U128`.

Comment (mneumann) [reply core/Integer.Format.savi:47]: We'll catch this by adding tests though I'd prefer when the compiler would complain when precision loss takes place (e.g. converting u64 -> u32).

**Comment by mneumann** [comment]
another approach could be to first convert the integer into a BCD (Binary Coded Decimal). a U64 can hold 16 decimal digits in packed format, so you'd need at least two U64 to hold all possible decimals of a U64.

The BCD can then very efficiently converted to ASCII digits.

The nice thing, you can convert the number from right to left, starting with the tenth, hundreds etc.

See this gist:

https://gist.github.com/mneumann/a34c1823f89cfd18cd49eeae89d46ccf

It does not require table lookups, but two loops.

**Comment by mneumann** [comment]
That would only use registers and only one "division by 10" per digit, instead of 3. Though, you probably would not notice that :)

**Pull request merged**